File: leetcode/algorithms/tree/n-ary-tree-postorder-traversal.java
/*
/*
    leetcode
    category: "Algorithms"
    problem title: "N-ary Tree Postorder Traversal"
*/


/* Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    
    private void populateListPostOrder(Node node, List<Integer> postOrdList){
        if(node==null){
            return;
        }
        for(int i=0;i<node.children.size();++i){
            populateListPostOrder(node.children.get(i),postOrdList);
        }
        postOrdList.add(node.val);
    }
    
    public List<Integer> postorder(Node root) {
        List<Integer> postOrdList = new ArrayList<Integer>();
        
        populateListPostOrder(root,postOrdList);
        
        return postOrdList;
        
        
    }
}

File: leetcode/algorithms/tree/n-ary-tree-postorder-traversal.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */

func addToPostorder(root *Node, postorderArr *[]int) {
	if root == nil {
		return
	}
	for _, childNode := range root.Children {

		addToPostorder(childNode, postorderArr)
	}

	*postorderArr = append(*postorderArr, root.Val)

}

func postorder(root *Node) []int {
	var arr []int
	var postorderArr *[]int = &arr

	addToPostorder(root, postorderArr)

	return *postorderArr
}